File: python/reorder-list.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        def find_mid(head):
            fast = slow = head
            while fast.next and fast.next.next:
                fast = fast.next.next
                slow = slow.next
            return slow
        
        def reverse(head):
            first, second = None, head
            while second:
                third = second.next
                second.next = first
                first, second = second, third
            return first
        
        def weave(left, right):
            head = curr = left
            while right:
                curr_next = curr.next
                curr.next = right
                right, left = curr_next, right.next
                curr = curr.next
            return head
        
        if head.next == None: return head
        
        dummy_head = ListNode(0, head)
        mid = find_mid(dummy_head)
        
        left = head
        right = mid.next
        mid.next = None
        
        
        right = reverse(right)
        
        weave(left, right)

File: go/reorder_list.go
package leetcode

type ListNode struct {
	Val  int
	Next *ListNode
}

func reorderList(head *ListNode) {
	if head.Next == nil {
		return
	}

	dummyHead := ListNode{0, head}
	mid := findMid(&dummyHead)

	left := head
	right := mid.Next
	mid.Next = nil

	right = reverse(right)

	weave(left, right)
}

func findMid(head *ListNode) *ListNode {
	fast := head
	slow := head

	for fast.Next != nil && fast.Next.Next != nil {
		fast = fast.Next.Next
		slow = slow.Next
	}
	return slow
}

func reverse(head *ListNode) *ListNode {
	var first *ListNode = nil
	second := head

	for second != nil {
		third := second.Next
		second.Next = first
		first, second = second, third
	}
	return first
}

func weave(left, right *ListNode) {
	curr := left

	for right != nil {
		curr_next := curr.Next
		curr.Next = right
		left, right = right.Next, curr_next
		curr = curr.Next
	}
}
